File: src/lib/taskNotes.test.ts
import { describe, expect, it } from "vitest";
import {
  collectTaskLists,
  collectTaskNotePaths,
  filterTaskIndex,
  getTaskAttrs,
  isTaskInCompleted,
  parseTaskNote,
  serializeTaskNote,
  setTaskAttrs,
  taskCompletedFolder,
  taskIndexEntryFromNote,
  taskListFromPath,
  type TaskIndexEntry,
  type TaskNote,
} from "./taskNotes";
import type { TreeNode } from "./vaultApi";

const SAMPLE = `---
status: open
due: 2026-08-28
priority: 2
labels: [work, report]
created: 2026-08-27
---

# Send report

## Subtasks

- [ ] Draft numbers
- [ ] Ask review
  - [x] Nested check
- [x] Export PDF

## Comments

### 2026-08-27 14:02

Looks good:

![](.assets/shot.png)

### 2026-08-27 18:10

Sent to boss.
`;

describe("taskNotes parse/serialize", () => {
  it("parses frontmatter and body sections", () => {
    const note = parseTaskNote("Tasks/Work/send-report.md", SAMPLE);
    expect(note.title).toBe("Send report");
    expect(note.attrs).toEqual({
      status: "open",
      due: "2026-08-28",
      priority: 2,
      labels: ["work", "report"],
      created: "2026-08-27",
      id: "",
      parent: null,
    });
    expect(note.subtasks).toHaveLength(4);
    expect(note.subtasks[0]!.text).toBe("Draft numbers");
    expect(note.subtasks[1]!.text).toBe("Ask review");
    expect(note.subtasks[1]!.children).toHaveLength(0);
    expect(note.subtasks[2]!.text).toBe("Nested check");
    expect(note.subtasks[2]!.checked).toBe(true);
    expect(note.subtasks[3]!.checked).toBe(true);
    expect(note.comments).toHaveLength(2);
    expect(note.comments[0]!.at).toBe("2026-08-27 14:02");
    expect(note.comments[0]!.body).toContain(".assets/shot.png");
    expect(note.comments[1]!.body).toBe("Sent to boss.");
  });

  it("round-trips structured fields", () => {
    const note = parseTaskNote("Tasks/Work/send-report.md", SAMPLE);
    const again = parseTaskNote(note.path, serializeTaskNote(note));
    expect(again.title).toBe(note.title);
    expect(again.attrs).toEqual(note.attrs);
    expect(again.subtasks).toEqual(note.subtasks);
    expect(again.comments).toEqual(note.comments);
  });

  it("uses file stem when title heading is missing", () => {
    const md = `---
status: done
---

## Subtasks

- [x] Only subtask
`;
    const note = parseTaskNote("Tasks/Inbox/buy-milk.md", md);
    expect(note.title).toBe("buy-milk");
    expect(note.attrs.status).toBe("done");
  });

  it("patches attrs while preserving body", () => {
    const next = setTaskAttrs(SAMPLE, { status: "done", priority: 1 });
    expect(getTaskAttrs(next).status).toBe("done");
    expect(getTaskAttrs(next).priority).toBe(1);
    expect(getTaskAttrs(next).due).toBe("2026-08-28");
    expect(next).toContain("# Send report");
    expect(next).toContain("## Comments");
  });

  it("builds index entry with list; child counts come from enrich", () => {
    const note = parseTaskNote("Tasks/Work/send-report.md", SAMPLE);
    const entry = taskIndexEntryFromNote(note);
    expect(entry.list).toBe("Work");
    expect(entry.parent).toBeNull();
    // File-child counts are 0 until enrichTaskIndexChildren runs.
    expect(entry.subtaskTotal).toBe(0);
    expect(entry.subtaskDone).toBe(0);
    expect(entry.commentCount).toBe(2);
  });

  it("parses parent as task UUID", () => {
    const md = `---
id: a1b2c3d4-e5f6-7890-abcd-ef1234567890
status: open
parent: 7f3a2c1e-9b4d-4e2a-a1c0-1234567890ab
---

# Draft numbers
`;
    const note = parseTaskNote("Tasks/Work/draft.md", md);
    expect(note.attrs.id).toBe("a1b2c3d4-e5f6-7890-abcd-ef1234567890");
    expect(note.attrs.parent).toBe("7f3a2c1e-9b4d-4e2a-a1c0-1234567890ab");
    const again = parseTaskNote(note.path, serializeTaskNote(note));
    expect(again.attrs.parent).toBe("7f3a2c1e-9b4d-4e2a-a1c0-1234567890ab");
    expect(again.attrs.id).toBe("a1b2c3d4-e5f6-7890-abcd-ef1234567890");
  });

  it("keeps legacy path parent until identity migration", () => {
    const md = `---
status: open
parent: Tasks/Work/send-report.md
---

# Draft numbers
`;
    const note = parseTaskNote("Tasks/Work/draft.md", md);
    expect(note.attrs.parent).toBe("Tasks/Work/send-report.md");
  });
});

function entry(partial: Partial<TaskIndexEntry> & Pick<TaskIndexEntry, "path" | "title">): TaskIndexEntry {
  return {
    id: partial.id ?? `id-${partial.path}`,
    status: "open",
    due: null,
    priority: null,
    labels: [],
    created: null,
    parent: null,
    list: "Inbox",
    subtaskTotal: 0,
    subtaskDone: 0,
    commentCount: 0,
    subtasks: [],
    description: "",
    ...partial,
  };
}

describe("filterTaskIndex", () => {
  const today = "2026-08-28";
  const rows: TaskIndexEntry[] = [
    entry({
      path: "Tasks/Inbox/a.md",
      title: "Milk",
      list: "Inbox",
      due: today,
      priority: 2,
    }),
    entry({
      path: "Tasks/Work/b.md",
      title: "Report",
      list: "Work",
      due: today,
      priority: 1,
      labels: ["work"],
    }),
    entry({
      path: "Tasks/Work/c.md",
      title: "Later",
      list: "Work",
      due: "2026-09-01",
    }),
    entry({
      path: "Tasks/Inbox/d.md",
      title: "Done inbox",
      list: "Inbox",
      status: "done",
    }),
  ];

  it("filters inbox open tasks", () => {
    const out = filterTaskIndex(
      rows,
      "inbox",
      { query: "", list: "", priority: "", label: "", status: "open" },
      today,
    );
    expect(out.map((e) => e.path)).toEqual(["Tasks/Inbox/a.md"]);
  });

  it("preserves vault order in inbox (no priority re-sort)", () => {
    const ordered: TaskIndexEntry[] = [
      entry({
        path: "Tasks/Inbox/z.md",
        title: "Zebra",
        list: "Inbox",
        priority: 4,
      }),
      entry({
        path: "Tasks/Inbox/a.md",
        title: "Alpha",
        list: "Inbox",
        priority: 1,
      }),
    ];
    const out = filterTaskIndex(
      ordered,
      "inbox",
      { query: "", list: "", priority: "", label: "", status: "open" },
      today,
    );
    expect(out.map((e) => e.path)).toEqual([
      "Tasks/Inbox/z.md",
      "Tasks/Inbox/a.md",
    ]);
  });

  it("filters today by due date", () => {
    const out = filterTaskIndex(
      rows,
      "today",
      { query: "", list: "", priority: "", label: "", status: "open" },
      today,
    );
    expect(out.map((e) => e.path)).toEqual([
      "Tasks/Work/b.md",
      "Tasks/Inbox/a.md",
    ]);
  });

  it("applies label and list filters", () => {
    const out = filterTaskIndex(
      rows,
      "filters",
      {
        query: "",
        list: "Work",
        priority: "",
        label: "work",
        status: "open",
      },
      today,
    );
    expect(out.map((e) => e.path)).toEqual(["Tasks/Work/b.md"]);
  });

  it("ignores sticky list filter on Today", () => {
    const out = filterTaskIndex(
      rows,
      "today",
      {
        query: "",
        list: "Work",
        priority: "",
        label: "",
        status: "open",
      },
      today,
    );
    expect(out.map((e) => e.path)).toEqual([
      "Tasks/Work/b.md",
      "Tasks/Inbox/a.md",
    ]);
  });

  it("ignores sticky priority filter on Inbox", () => {
    const out = filterTaskIndex(
      rows,
      "inbox",
      {
        query: "",
        list: "",
        priority: 3,
        label: "",
        status: "open",
      },
      today,
    );
    expect(out.map((e) => e.path)).toEqual(["Tasks/Inbox/a.md"]);
  });
});

describe("serializeTaskNote minimal", () => {
  it("omits empty optional sections", () => {
    const note: Omit<TaskNote, "path"> = {
      title: "Quick",
      attrs: {
        status: "open",
        due: null,
        priority: null,
        labels: [],
        created: "2026-08-27",
        id: "11111111-1111-4111-8111-111111111111",
        parent: null,
      },
      description: "",
      subtasks: [],
      comments: [],
    };
    const md = serializeTaskNote(note);
    expect(md).toContain("status: open");
    expect(md).toContain("created: 2026-08-27");
    expect(md).not.toContain("due:");
    expect(md).not.toContain("## Subtasks");
    expect(md).not.toContain("## Comments");
    expect(md).toContain("# Quick");
  });
});

describe("completed archive paths", () => {
  it("resolves list and completed folder helpers", () => {
    expect(taskListFromPath("Tasks/Work/send-report.md")).toBe("Work");
    expect(taskListFromPath("Tasks/Inbox/completed/old.md")).toBe("Inbox");
    expect(taskListFromPath("Tasks/completed/orphan.md")).toBe("");
    expect(taskCompletedFolder("Work")).toBe("Tasks/Work/completed");
    expect(isTaskInCompleted("Tasks/Work/completed/old.md")).toBe(true);
    expect(isTaskInCompleted("Tasks/Work/send-report.md")).toBe(false);
    expect(isTaskInCompleted("Tasks/completed/x.md")).toBe(false);
  });

  it("skips completed folders when collecting active task paths and lists", () => {
    const tree: TreeNode = {
      name: "",
      path: "",
      isDir: true,
      children: [
        {
          name: "Tasks",
          path: "Tasks",
          isDir: true,
          children: [
            {
              name: "Inbox",
              path: "Tasks/Inbox",
              isDir: true,
              children: [
                {
                  name: "open.md",
                  path: "Tasks/Inbox/open.md",
                  isDir: false,
                  children: [],
                },
                {
                  name: "completed",
                  path: "Tasks/Inbox/completed",
                  isDir: true,
                  children: [
                    {
                      name: "done.md",
                      path: "Tasks/Inbox/completed/done.md",
                      isDir: false,
                      children: [],
                    },
                  ],
                },
              ],
            },
            {
              name: "completed",
              path: "Tasks/completed",
              isDir: true,
              children: [],
            },
            {
              name: "Work",
              path: "Tasks/Work",
              isDir: true,
              children: [],
            },
          ],
        },
      ],
    };
    expect(collectTaskNotePaths(tree)).toEqual(["Tasks/Inbox/open.md"]);
    expect(collectTaskLists(tree)).toEqual(["Inbox", "Work"]);
  });

  it("serializes an appended comment block", () => {
    const note = parseTaskNote(
      "Tasks/Inbox/hello.md",
      `---
id: 11111111-1111-4111-8111-111111111111
status: open
---

# Hello
`,
    );
    const withComment = {
      ...note,
      comments: [{ at: "2026-08-30 12:00", body: "Hi" }],
    };
    const out = serializeTaskNote(withComment);
    expect(out).toContain("## Comments");
    expect(out).toContain("### 2026-08-30 12:00");
    expect(out).toContain("Hi");
  });
});

File: android-tasks/tasknotes/src/test/java/com/markspace/tasknotes/TaskNotesTest.kt
package com.markspace.tasknotes

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

class TaskNotesTest {

    private fun fixture(name: String): String {
        val prop = System.getProperty("tasknotes.fixtures")
        val dir = if (prop != null) File(prop) else {
            // Fallback: walk up from user.dir
            var d = File(System.getProperty("user.dir"))
            while (d != null) {
                val cand = File(d, "fixtures/task-notes")
                if (cand.isDirectory) return File(cand, name).readText()
                d = d.parentFile
            }
            error("fixtures/task-notes not found")
        }
        return File(dir, name).readText()
    }

    @Test
    fun parsesFrontmatterAndBody() {
        val note = parseTaskNote("Tasks/Work/send-report.md", fixture("send-report.md"))
        assertEquals("Send report", note.title)
        assertEquals("open", note.attrs.status)
        assertEquals("2026-08-28", note.attrs.due)
        assertEquals(2, note.attrs.priority)
        assertEquals(listOf("work", "report"), note.attrs.labels)
        assertEquals("2026-08-27", note.attrs.created)
        assertEquals(4, note.subtasks.size)
        assertEquals("Draft numbers", note.subtasks[0].text)
        assertEquals("Nested check", note.subtasks[2].text)
        assertTrue(note.subtasks[2].checked)
        assertEquals(2, note.comments.size)
        assertEquals("2026-08-27 14:02", note.comments[0].at)
        assertTrue(note.comments[0].body.contains(".assets/shot.png"))
    }

    @Test
    fun roundTripsStructuredFields() {
        val note = parseTaskNote("Tasks/Work/send-report.md", fixture("send-report.md"))
        val again = parseTaskNote(note.path, serializeTaskNote(note))
        assertEquals(note.title, again.title)
        assertEquals(note.attrs, again.attrs)
        assertEquals(note.subtasks, again.subtasks)
        assertEquals(note.comments, again.comments)
    }

    @Test
    fun usesFileStemWhenTitleMissing() {
        val note = parseTaskNote("Tasks/Inbox/buy-milk.md", fixture("no-title.md"))
        assertEquals("buy-milk", note.title)
        assertEquals("done", note.attrs.status)
    }

    @Test
    fun parsesParentUuid() {
        val note = parseTaskNote("Tasks/Work/draft.md", fixture("child-with-parent.md"))
        assertEquals("a1b2c3d4-e5f6-7890-abcd-ef1234567890", note.attrs.id)
        assertEquals("7f3a2c1e-9b4d-4e2a-a1c0-1234567890ab", note.attrs.parent)
        val again = parseTaskNote(note.path, serializeTaskNote(note))
        assertEquals(note.attrs.parent, again.attrs.parent)
    }

    @Test
    fun minimalSerializeOmitsEmptyOptionals() {
        val note = parseTaskNote("Tasks/Inbox/quick.md", fixture("minimal.md"))
        val md = serializeTaskNote(note)
        assertTrue(md.contains("status: open"))
        assertTrue(md.contains("created: 2026-08-27"))
        assertFalse(md.contains("due:"))
        assertFalse(md.contains("## Subtasks"))
        assertFalse(md.contains("## Comments"))
        assertTrue(md.contains("# Quick"))
    }

    @Test
    fun indexEntryListAndComments() {
        val note = parseTaskNote("Tasks/Work/send-report.md", fixture("send-report.md"))
        val entry = taskIndexEntryFromNote(note)
        assertEquals("Work", entry.list)
        assertNull(entry.parent)
        assertEquals(0, entry.subtaskTotal)
        assertEquals(2, entry.commentCount)
    }

    @Test
    fun filterInboxTodayUpcoming() {
        val today = "2026-08-28"
        fun entry(
            path: String,
            title: String,
            list: String,
            due: String? = null,
            priority: Int? = null,
            status: String = "open",
            labels: List<String> = emptyList(),
        ) = TaskIndexEntry(
            path = path,
            id = "id-$path",
            title = title,
            status = status,
            due = due,
            priority = priority,
            labels = labels,
            created = null,
            parent = null,
            list = list,
        )
        val rows = listOf(
            entry("Tasks/Inbox/a.md", "Milk", "Inbox", today, 2),
            entry("Tasks/Work/b.md", "Report", "Work", today, 1, labels = listOf("work")),
            entry("Tasks/Work/c.md", "Later", "Work", "2026-09-01"),
            entry("Tasks/Inbox/d.md", "Done inbox", "Inbox", status = "done"),
        )
        assertEquals(
            listOf("Tasks/Inbox/a.md"),
            filterTaskIndex(rows, TasksViewId.Inbox, TasksFilters(status = "open"), today).map { it.path },
        )
        assertEquals(
            listOf("Tasks/Work/b.md", "Tasks/Inbox/a.md"),
            filterTaskIndex(rows, TasksViewId.Today, TasksFilters(), today).map { it.path },
        )
        assertEquals(
            listOf("Tasks/Work/c.md"),
            filterTaskIndex(rows, TasksViewId.Upcoming, TasksFilters(), today).map { it.path },
        )
        assertEquals(
            listOf("Tasks/Work/b.md"),
            filterTaskIndex(
                rows,
                TasksViewId.Filters,
                TasksFilters(list = "Work", label = "work", status = "open"),
                today,
            ).map { it.path },
        )
    }

    @Test
    fun pathHelpers() {
        assertEquals("Work", taskListFromPath("Tasks/Work/send-report.md"))
        assertEquals("Inbox", taskListFromPath("Tasks/Inbox/completed/old.md"))
        assertEquals("", taskListFromPath("Tasks/completed/orphan.md"))
        assertEquals("Tasks/Work/completed", taskCompletedFolder("Work"))
        assertTrue(isTaskInCompleted("Tasks/Work/completed/old.md"))
        assertFalse(isTaskInCompleted("Tasks/Work/send-report.md"))
    }

    @Test
    fun repositoryRoundTrip() {
        val tmp = createTempDir(prefix = "tasknotes-repo-")
        try {
            val repo = TaskRepository(tmp)
            val note = repo.createTask("Buy milk", list = "Inbox", due = "2026-08-28", priority = 1)
            assertTrue(note.attrs.id.isNotEmpty())
            assertTrue(File(tmp, note.path).exists())
            repo.setStatus(note.path, "done")
            assertEquals("done", repo.loadNote(note.path).attrs.status)
            repo.appendComment(note.path, "Purchased")
            assertEquals(1, repo.loadNote(note.path).comments.size)
            val child = repo.createTask("Check price", parent = note.attrs.id)
            val index = repo.loadIndex()
            val parent = index.first { it.path == note.path }
            assertEquals(1, parent.subtaskTotal)
            assertEquals(0, parent.subtaskDone)
            assertEquals(note.attrs.id, index.first { it.path == child.path }.parent)
        } finally {
            tmp.deleteRecursively()
        }
    }
}
